Release 1.3.0

gioamato opened the pull request.

gioamato committed:
Add option to track last parameters

gioamato committed:
Add new option documentation

gioamato committed:
Fix cookie value on empty or missing parameter

gioamato committed:
Build version 1.3.0

medius commented on README.md:
```diff
@@ -119,6 +119,7 @@ If you would like to customize how fields get added to your form, following opti
     affiliate: "IAFFILIATE"
   }; // Maps the initial affiliate parameter to IAFFILIATE form field
 
+  _uf.track_last_params          = true; // Enable the tracking of last parameters even if null
```
Perhaps using `track_null_params` with comment _**Default: false. Use empty values if UTM not present on the same page as the form**_ can clarify this.

gioamato replied:
As i'm adding a `current_session` check maybe the comment should be **_Default: false. Use null values if UTM not present in the same session as the form submit_**, do you agree?

medius replied:
Rethinking about this (based on your `current_session` changes), it seems that the correct setting might be `reset_params_at_session_end`. 

Also, the comment could be _**Default: false. Use empty string values if UTM not present in the same session as the form submit**_

medius commented on lib/utm_cookie.coffee:
```diff
@@ -70,9 +74,10 @@ class UtmCookie
+      if value.length > 0 then value else null
     else
-      ''
+      null
```
Can you please elaborate the advantage of `null` over an empty string?

gioamato replied:
It's required for initial parameters (otherwise they will be overwritten) and also for the "lastest-even-null" logic.

medius replied:
To be honest, I'm still not sure about this. Since this method returns value to be written into the cookies, they'll be string with value "null" which seems incorrect. If there's no value, the cookie value should be an empty string.

Are you seeing something different for this case?

gioamato replied:
It is known that empty cookie values may not behave the same way on all browsers.

In our specific case an empty value will also fail the existingValue check inside writeCookieOnce method.

I also don’t like the “null” value. So what would you suggest?

medius replied:
That's a good point. In that case `writeCookie` needs to be updated to write only when `value` is a non-empty string. Similarly `readCookie` should return an empty string if no cookie is found. 

This might be a broader change that might be out of scope of your current goal. I can make these changes when I get the opportunity and also incorporate the changes proposed in this PR.

medius replied:
On second thought, `readCookie` should return `null`, but when using the value to fill the fields, it could be an empty string when the value is `null`. This way the role of these methods is clearer.

What do you think?

gioamato replied:
No problem, i can make the changes. But what about my first sentence? Do you still prefer to keep going with an empty cookie value?

medius replied:
No, we should change that. You are right that the cookie value shouldn't be empty. But returning `null` here might not work since `writeUtmCookieFromParams` and other such methods write the cookie without checking if the value is null or not. So `null` value will be written as "null" string in the cookie (at least that's what I see in my basic testing).

gioamato replied:
Cookie values are actually just strings. So a "real" null value, int, etc. cannot be stored inside cookies. So i think we should only choose a conventional value to identify a cookie without arbitrary data. Initially I thought about `null` or `0`, and proceeded with null.

Also cookie data should be encoded (like Base64) as for [RFC 6265](https://tools.ietf.org/html/rfc6265)

medius replied:
Yes, I do realize that a real null value cannot be stored in cookies. My main concern with the changes is that if say `utm_campaign` is not present in the URL, then the form will be populated with UTM campaign as "null". It should have it as '' (empty string).

Do you agree that with these changes, "null" string will be written in cookies and the form fields will have "null" for any UTM value that is missing in the URL? If not, then maybe I have misunderstood something.

gioamato replied:
I absolutely agree with you. The form fields are actually populated with `"null"` strings. They should be filled with `""` strings instead.

This PR is not ready to be merged. I'm just waiting your confirmations about the changes to do. As i said, i can make all the changes. So, let's choose our "conventions" and desired state and i will then start working on it.

medius replied:
Yes, that works for me. This is the only concern I had regarding these changes.

gioamato committed:
Fix option naming and add 'curent_session' check

gioamato commented:
• Fixed option naming, now it's `track_null_params`
• Added a check for `current_session` value to keep the paramaters during the session

Waiting for a confirmation about docs.

medius commented:
To clarify, how about these requirements for this PR?

- Keep the same behavior, i.e. use '' (empty string) in form field if a UTM value is not available.
- Reset UTM values at the end of session if `reset_params_at_session_end` is set to true.

gioamato committed:
Fix form behavior and update docs

gioamato commented on lib/utm_form.coffee:
```diff
@@ -100,7 +100,7 @@ class UtmForm
     fieldEl.type = "hidden"
     fieldEl.name = fieldName
-    fieldEl.value = if @_decodeURIs then decodeURIComponent(fieldValue) else fieldValue
+    fieldEl.value = if fieldValue then if @_decodeURIs then decodeURIComponent(fieldValue) else fieldValue else fieldValue
```
`decodeURIComponent` of `null` returns `'null'` (a string with null value).

This caused the form field to have `value='null'` attribute instead of `value=''`

medius replied:
Instead of the nested logic, how about this?

```javascript
value = fieldValue || ''
fieldEl.value = if @_decodeURIs then decodeURIComponent(value) else value
```

This makes it more readable and easy to understand.

gioamato commented:
> To clarify, how about these requirements for this PR?
> 
> * Keep the same behavior, i.e. use '' (empty string) in form field if a UTM value is not available.
> * Reset UTM values at the end of session if `reset_params_at_session_end` is set to true.

Done. Let me know if everything is ok for you.

medius commented on lib/utm_cookie.coffee:
```diff
@@ -56,7 +61,9 @@ class UtmCookie
-        return c.substring(nameEQ.length, c.length)
+        value = c.substring(nameEQ.length, c.length)
+        if value != 'null'
+          return value
```
Rather than storing and managing 'null' string in cookies, we can choose not to store any null values.

If we update `writeCookie` method as follow, no null or empty string will be written.

```javascript
 writeCookie: (name, value) ->
    if !!value
       @createCookie name, value, @_cookieExpiryDays, null, @_domain, @_secure
    return
```

gioamato replied:
In this way the initial parameters and the new feature `reset_params_at_session_end` will not work, breaking also the release 1.2.0.

medius replied:
Can you please elaborate why won't those work? Also, in what ways does it break 1.2.0 release? I'm sorry, I'm a bit confused here.

I am hesitant to store 'null' strings in cookie and then ignore them while reading. It creates more cruft that need to be kept track of.

gioamato replied:
The `initial_utm_params` and `additional_initial_params_map` are based on the `writeCookieOnce` method that checks for cookie existence before writing it.

medius replied:
Yes, if nothing is written, nothing will be read. 

To be honest, I'm having a hard time following these changes which is a bad sign for something that's supposed to be simple. Perhaps I'm not thinking straight and need to revisit these when I have more spare time. Since you have your own fork, this doesn't block you from using these changes for your own use.

gioamato replied:
This is super simple. As i said some methods are based on the `writeCookieOnce` one. This method will try to read the cookie before writing it. So if we don't write anything in the cookie it will always write the cookie. So we must store `null`, `0` or something in the cookie to keep the blank values, like in `initial_utm_params`.

If you don't want to store null values in the cookies then another way could be to use the visits counter and the current section cookie to "know" when the `writeCookieOnce` method should and should not write the cookie.

Let me know what you prefer. As i said, i have some spare time to do the job, no problem.

medius commented on lib/utm_cookie.coffee:
```diff
@@ -70,9 +77,10 @@ class UtmCookie
+      if value.length > 0 then value else null
     else
-      ''
+      null
```
Perhaps this code can be simplified as

```javascript
if results
  value = decodeURIComponent results[1].replace(/\+/g, ' ')
value || null
```
